### ISSUE TITLE
Decompose 'Scalability Testing' and provide estimation time for each task

### ISSUE THREAD

**Issue by Bolodya1997** [networkservicemesh/sdk#822]
## Description
The result of this issue should be a list of issues with approximate estimations related to _Scalability Testing_.

## Subtasks
* NSM testing - https://github.com/networkservicemesh/deployments-k8s/issues/1015
 Estimation: 6d
* NSMgr testing - https://github.com/networkservicemesh/deployments-k8s/issues/1016
 Estimation: 8d
* Registry testing - https://github.com/networkservicemesh/deployments-k8s/issues/1017
 Estimation: 5d
* Forwarder testing - https://github.com/networkservicemesh/deployments-k8s/issues/1018
 Estimation: 3d
* Endpoint testing - https://github.com/networkservicemesh/deployments-k8s/issues/1019
 Estimation: 3d
* Client testing - https://github.com/networkservicemesh/deployments-k8s/issues/1020
 Estimation: 1d

## Estimation
`26d`

**Comment by Bolodya1997** [comment]
For this task we need:
1. Multiple clients testing (5, 10, 20, 100, 500) for cluster on (local, remote) cases.
 Estimation: 5h (~30 minutes for each case).
2. Multiple endpoints testing (5, 10, 20, 100, 500) for cluster on (local, remote) cases.
 Estimation: 5h (~30 minutes for each case).
3. Multiple forwarders testing (5, 10) for single node on (local, remote) cases.
 Estimation: 3h (~40 minutes for each case).
4. Mixed testing with multiple (5, 10, 20, 100, 500) clients, endpoints on (local, remote) cases.
 Estimation: 5h (~30 minutes for each case).
5. Bug fixes.
 Estimation: 12 days (~3 days for each of 1,2,3,4 testings).

**Comment by denis-tingaikin** [comment]
I think it is blocked by https://github.com/networkservicemesh/sdk/issues/790

**Comment by denis-tingaikin** [comment]
@edwarnicke Can we implement #790 before starting this?

**Comment by edwarnicke** [comment]
> For this task we need:
> 
> 1. Multiple clients testing (5, 10, 20, 100, 500) for cluster on (local, remote) cases.
>    Estimation: 6h (~30m for each case + ~1h for creating automation).
> 2. Multiple endpoints testing (5, 10, 20, 100, 500) for cluster on (local, remote) cases.
>    Estimation: 6h (~30m for each case + ~1h for creating automation).
> 3. Multiple forwarders testing (5, 10) for single node on (local, remote) cases.
>    Estimation: 4h (~40m for each case + ~1h for creating automation).
> 4. Mixed testing with multiple (5, 10, 20, 100, 500) clients, endpoints on (local, remote) cases.
>    Estimation: 6h (~30m for each case + ~1h for creating automation).
> 5. Bug fixes.
>    Estimation: 64h (~16h for each of 1, 2, 3, 4).
> 
> Total: 86h

I this an estimate of runtime of each case, or the time to do the work of setting each case up?

**Comment by edwarnicke** [comment]
@Bolodya1997 Also... is this (client, endpoint, forwarder) the right breakdown for testing scalability?

For example... if you look at the scalability characteristics of the system, you have:
1.  Scalability of the registry (both in terms of size and rate of query/update)
2. Scalability of the NSMgr per Node
3. Scalability of each Forwarder per Node
4. Scalability of a particular NSE (ie, how many clients can it handle, and how quickly can it handle them coming and going)
5. Scalability of the NSC client (ie, how many NSEs can it attach to and how quickly)

**Comment by denis-tingaikin** [comment]
Another way: we can ignore 
> Multiple forwarders testing
for  a while
And start with cases that @edwarnicke  suggested.

Thoughts?

**Comment by edwarnicke** [comment]
@Bolodya1997 From some discussions I had with @denis-tingaikin 

There are some external limits to certain kinds of scale:
1. K8s Nodes tend to cap out at 100 Pods
2. Linux has a limit of 1024 kernel interfaces

which mean we cannot do things like, for example, running 500 NSCs on a single Node.

But you could look at scale scenarios like:

1. Start 100 Pods at once on a Single Node each of which is an NSC client for two NSes
2. Start 100 NSEs on a Node0.  Start 100 NSCs each on Node[1-10] each trying to connect o two NSes represented by those NSEs.  That would be 2k connections coming into Node0's NSMgr and Forwarder... but since we would naturally run the NSEs with memif mechanisms... we wouldn't hit the kernel interface limit.

Those are just a couple of examples.  I'm sure we can think of others :)

**Comment by edwarnicke** [comment]
> Another way: we can ignore
> 
> > Multiple forwarders testing
> 
> for a while
> And start with cases that @edwarnicke suggested.
> 
> Thoughts?

Obviously... we can't test multi-forwarder scaling till we have it working properly.

**Comment by Bolodya1997** [comment]
> I this an estimate of runtime of each case, or the time to do the work of setting each case up?

An estimate time to do the work, but I guess it should be reevaluated.

**Comment by edwarnicke** [comment]
@Bolodya1997 You did a great first pass of thinking it through.  Like most first passes, you just need a second pass :)

**Comment by Bolodya1997** [comment]
## Registry
**Use cases:**
1. NSMgr registers Endpoint on Registry.
2. NSMgr updates Endpoint on Registry.
3. NSMgr unregisters Endpoint from Registry.
4. NSMgr creates a find request to Registry.
5. NSMgr creates a watching find request to Registry.
6. Registry unregisters expired Endpoint on Registry.

**Test scenarios:**
1. NSMgr registers Endpoint on Registry:
   * **Setup**
      1. Start Registry.
   * **Test**
      1. Start _N_ fake NSMgrs each registering _E_ Endpoints.
2. NSMgr updates Endpoint on Registry:
   * **Setup**
      1. Start Registry.
      2. Start _N_ fake NSMgrs each registering _E_ Endpoints with _T_ update time.
   * **Test**
      1. Wait _T_ time for updates.
3. NSMgr unregisters Endpoint from Registry:
   * **Setup**
      1. Start Registry.
      2. Start _N_ fake NSMgrs each registering _E_ Endpoints.
   * **Test**
      1. Unregister endpoints(?).
4. NSMgr creates a find request to Registry:
   * **Setup**
      1. Start Registry.
      2. Start _N_ fake NSMgrs each registering _E_ Endpoints.
   * **Test**
      1. Create _F_ find requests from each NSMgr to Registry.
5. NSMgr creates a watching find request to Registry:
   * **Setup**
      1. Start Registry.
      2. Start _N_ fake NSMgrs each registering _E_ Endpoints.
   * **Test**
      1. Create _W_ watching find requests from each NSMgr to Registry.
6. Registry unregisters expired Endpoint on Registry:
   * **Setup**
      1. Start Registry.
      2. Start _N_ fake NSMgrs each registering _E_ Endpoints with _T_ expire time.
   * **Test**
      1. Wait _T_ time for unregisters.

**Comment by Bolodya1997** [comment]
## NSMgr

**Use cases:**
1. Endpoint registers on NSMgr.
2. Endpoint updates itself on NSMgr.
3. Endpoint unregisters from NSMgr.
4. NSMgr unregisters expired Endpoint from itself.
5. Forwarder registers on NSMgr.
6. Forwarder updates itself on NSMgr.
7. Forwarder unregisters from NSMgr.
8. NSMgr unregisters expired Forwarder from itself.
9. Client sends request to NSMgr.
10. Client sends refresh request to NSMgr.
11. Client closes connection on NSMgr.

**Test scenarios:**
1. Endpoint registers on NSMgr:
   * **Setup**
      1. Start fake Registry (just returning OK to all Register/Unregister events).
      2. Start NSMgr.
   * **Test**
      1. Start _E_ Endpoints each registering itself on NSMgr.
2. Endpoint updates itself on NSMgr:
   * **Setup**
      1. Start fake Registry (just returning OK to all Register/Unregister events).
      2. Start NSMgr.
      3. Start _E_ Endpoints each registering itself on NSMgr with _T_ update time.
   * **Test**
      1. Wait _T_ time for updates.
3. Endpoint unregisters from NSMgr:
   * **Setup**
      1. Start fake Registry (just returning OK to all Register/Unregister events).
      2. Start NSMgr.
      3. Start _E_ Endpoints each registering itself on NSMgr.
   * **Test**
      1. Unregister Endpoints.
4. NSMgr unregisters expired Endpoint from itself:
   * **Setup**
      1. Start fake Registry (just returning OK to all Register/Unregister events).
      2. Start NSMgr.
      3. Start _E_ Endpoints each registering itself on NSMgr with _T_ expiration time.
   * **Test**
      1. Wait _T_ time for unregisters.
5. Forwarder registers on NSMgr:
   * (1)
6. Forwarder updates itself on NSMgr:
   * (2)
7. Forwarder unregisters from NSMgr:
   * (3)
8. NSMgr unregisters expired Forwarder from itself:
   * (4)
9. Client sends request to NSMgr:
   * **Setup**
      1. Start NSMgr.
      2. Start _E_ Endpoints each registering itself on NSMgr.
   * **Test**
      1. Start _C_ Clients each requesting for _E_ network services (Endpoints).
10. Client sends refresh request to NSMgr:
    * **Setup**
      1. Start NSMgr.
      2. Start _E_ Endpoints each registering itself on NSMgr.
      3. Start _C_ Clients each requesting for _E_ Endpoints with _T_ refresh time.
    * **Test**
      1. Wait _T_ time for refreshes.
11. Client closes connection on NSMgr.
    * **Setup**
      1. Start NSMgr.
      2. Start _E_ Endpoints each registering itself on NSMgr.
      3. Start _C_ Clients each requesting for _E_ Endpoints.
    * **Test**
       1. Cancel Clients.
12. NSMgr closes expired connection with Client.
    * **Setup**
      1. Start NSMgr.
      2. Start _E_ Endpoints each registering itself on NSMgr.
      3. Start _C_ Clients each requesting for _E_ Endpoints with _T_ expiration time.
    * **Test**
      1. Wait _T_ time for closes.

**Comment by Bolodya1997** [comment]
## Forwarder

**Use cases:**

1. NSMgr sends request to Forwarder.
2. NSMgr sends refresh request to Forwarder.
3. NSMgr closes connection on Forwarder.
4. Forwarder closes expired connection with NSMgr.

**Test scenarios:**

1. NSMgr sends request to Forwarder.
   * **Setup**
      1. Start fake NSMgr (handling both Client/Remote and Endpoint/Remote sides).
      2. Start Forwarder.
   * **Test**
      1. Request _C_ connections with _Cm_ Client mechanism and _Em_ Endpoint mechanism.
2. NSMgr sends refresh request to Forwarder.
   * **Setup**
      1. Start fake NSMgr (handling both Client/Remote and Endpoint/Remote sides).
      2. Start Forwarder.
      3. Request _C_ connections with _Cm_ Client mechanism and _Em_ Endpoint mechanism with _T_ refresh time.
   * **Test**
      1. Wait _T_ time for refreshes.
3. NSMgr closes connection on Forwarder.
   * **Setup**
      1. Start fake NSMgr (handling both Client/Remote and Endpoint/Remote sides).
      2. Start Forwarder.
      3. Request _C_ connections with _Cm_ Client mechanism and _Em_ Endpoint mechanism.
   * **Test**
      1. Close connections.
4. Forwarder closes expired connection with NSMgr.
   * **Setup**
      1. Start fake NSMgr (handling both Client/Remote and Endpoint/Remote sides).
      2. Start Forwarder.
      3. Request _C_ connections with _Cm_ Client mechanism and _Em_ Endpoint mechanism with _T_ expiration time.
   * **Test**
      1. Wait _T_ time for closes.

**Comment by Bolodya1997** [comment]
## Endpoint

**Use cases:**

1. NSMgr sends request to Endpoint.
2. NSMgr sends refresh request to Endpoint.
3. NSMgr closes connection on Endpoint.
4. Endpoint closes expired connection with NSMgr.

**Test scenarios:**

1. NSMgr sends request to Endpoint:
   * **Setup**
      1. Start fake NSMgr.
      2. Start Endpoint.
   * **Test**
      1. Request _C_ connections.
2. NSMgr sends refresh request to Endpoint:
   * **Setup**
      1. Start fake NSMgr.
      2. Start Endpoint.
      3. Request _C_ connections.
   * **Test**
      1. Wait _T_ time for refreshes.
3. NSMgr closes connection on Endpoint:
   * **Setup**
      1. Start fake NSMgr.
      2. Start Endpoint.
      3. Request _C_ connections.
   * **Test**
      1. Close connections.
4. Endpoint closes expired connection with NSMgr:
   * **Setup**
      1. Start fake NSMgr.
      2. Start Endpoint.
      3. Request _C_ connections.
   * **Test**
      1. Wait _T_ time for closes.

**Comment by Bolodya1997** [comment]
## Client

**Use cases:**

1. Client sends request to NSMgr.
2. Client sends refresh request to NSMgr.

**Test scenarios:**

1. Client sends request to NSMgr:
   * **Setup**
      1. Start fake NSMgr.
   * **Test**
      1. Start Client requesting _C_ connections.
2. Client sends refresh request to NSMgr:
   * **Setup**
      1. Start fake NSMgr.
      2. Start Client requesting _C_ connections with _T_ refresh time.
      3. Wait some time before _T_.
   * **Test**
      1. Wait time until _T_ for refreshes.

**Comment by Bolodya1997** [comment]
For all mentioned **Tests** we need to measure:
1. Memory usage during the test.
2. CPU usage during the test.
3. Time of the test.

**Comment by Bolodya1997** [comment]
@edwarnicke 
Do we need tests for the whole NSM system? Probably something like this with measuring memory, CPU, time for all NSM components.

> * Multiple clients testing for cluster on (local, remote) cases.
> * Multiple endpoints testing for cluster on (local, remote) cases.
> * Mixed testing with multiple clients, endpoints on (local, remote) cases.

**Comment by edwarnicke** [comment]
@Bolodya1997 So this gets to an interesting question... we basically have per-component scalability and systemic scalability.

You've laid out some good per-component scalability testing.  That's awesome, and probably the place to start.  I do expect we will find new sorts of scalability issues when we get to systemic scalability.

I like where you are going as well with measuring things: cpu, memory, latency, etc.

**Comment by denis-tingaikin** [comment]
@edwarnicke here also we got a few questions:
1. Can we use in scalability testing special 'mock' applications or we could use only real applications?
2. If we can use 'mock' application then where we can store them?
3. Should we put scalability tests into deployments-k8s as an example?
4. Should we run scalability tests on CI?

**Comment by edwarnicke** [comment]
> @edwarnicke here also we got a few questions:
> 
> 1. Can we use in scalability testing special 'mock' applications or we could use only real applications?

For component wrapping Mocks around them is fine.  For systemic scalability you should use the real system... no mocks.

> 2. If we can use 'mock' applications then where we can store them?
Say more about what you mean by 'mock' applications

> 3. Should we put scalability tests into deployments-k8s as examples?
For systemic scalability tests... maybe?  Part of the question is whether or not we need to move away for MD Driven testing for some of the scalability tests or not.  The other question becomes one of organization for deployments-k8s as we want to continue to have it be easy for users to navigate to find the examples they want to try.

> 4. Should we run scalability tests on CI?
That depends on how expensive in time and resources they are.  If they are not to expensive in time and resources then 'why not'... but my guess is they are things that will be expensive in time and resources and so perhaps run at intervals in time rather than run commit by commit.

> 5. Where we can publish the result of scalability testing?
That's a good question... perhaps output should push a PR to the 'site' repo ? :)

**Comment by denis-tingaikin** [comment]
>> 2. If we can use 'mock' applications then where we can store them?
> Say more about what you mean by 'mock' applications

@Bolodya1997 Could you explain your idea with wrapping Mocks around the testing application that we discussed locally?

> > 3. Should we put scalability tests into deployments-k8s as examples?
> The other question becomes one of organization for deployments-k8s as we want to continue to have it be easy for users to navigate to find the examples they want to try.

I think we can try to start with MD examples and if it will be too hard then we can consider something else (something like using `k8s client go` in this kind of tests).


> > 4. Should we run scalability tests on CI?
>  but my guess is they are things that will be expensive in time and resources and so perhaps run at intervals in time rather than run commit by commit.

Sounds good to me. We can schedule them per one or two weeks vis GitHub actions :)


>>5. Where we can publish the result of scalability testing?

> That's a good question... perhaps output should push a PR to the 'site' repo ? :)

I like this idea :)

**Comment by edwarnicke** [comment]
> I think we can try to start with MD examples and if it will be too hard then we can consider something else (something like using `k8s client go` in this kind of tests).

Seems wise :)

**Comment by Bolodya1997** [comment]
> > Say more about what you mean by 'mock' applications
> 
> @Bolodya1997 Could you explain your idea with wrapping Mocks around the testing application that we discussed locally?

* If we are testing NSMgr capabilities of registering Endpoints we don't actually want _Registry_ to store these Endpoints (because it can affect on test results) but we probably need some _fake Registry_ instantly returning OK for each request, because communication with Registry is a part of the NSMgr registration process.
* If we are testing Registry, we don't want to create _N NSMgrs with E endpoints_ on each, but it would be OK to create _N fake NSMgrs_ behaving as those real world NSMgrs + Endpoints.

So there will be a need to create 1-2 testing applications for each component and to store them somewhere.

**Comment by edwarnicke** [comment]
@Bolodya1997 Its OK to do 'unit test level' scalability with 'fake' things... but integration level scalability needs to be tested with real thing.

**Comment by Bolodya1997** [comment]
> @Bolodya1997 Its OK to do 'unit test level' scalability with 'fake' things... but integration level scalability needs to be tested with real thing.

@edwarnicke Yep, it is exactly what we want to do :)

So finally the question is: _if we want to use some fake per-test CMD applications, would it be OK to store them not in some separate cmd- repositories, but in the same repository with the scalability tests?_

**Comment by edwarnicke** [comment]
@Bolodya1997 Are these running as separate processes (ie, real executables with main.go files)?

**Comment by Bolodya1997** [comment]
> @Bolodya1997 Are these running as separate processes (ie, real executables with main.go files)?

Yep, I think so, because scalability should be tested in real k8s environment I suppose.

**Comment by edwarnicke** [comment]
@Bolodya1997 So what I'm confused about is that clearly for running scalability in the real K8s env, we want separate executables, in the form of separate cmd-* repos.  What I'm not clear on is if you are instead asking we put those into the sdk repo so we can run fake per-test applications for sandbox...

**Comment by Bolodya1997** [comment]
@edwarnicke No, I mean creating some single repository for all scalability testing stuff instead of creating lots of `cmd-fake-` repositories.

**Comment by edwarnicke** [comment]
@Bolodya1997 How many do you think there will be?  I expected we would mostly be reusing existing applications, only more of them...

**Comment by denis-tingaikin** [comment]
@edwarnicke , @Bolodya1997 

Can we start with system scalability testing that includes all real applications?

For component scalability testing we could use the sandbox as mock and add the testing on cmd level.

What do you think? Please share your thoughts.

**Comment by edwarnicke** [comment]
@denis-tingaikin Yes, we can start with system scalability and real applications.

**Comment by Bolodya1997** [comment]
Yes, I guess so.

**Comment by Bolodya1997** [comment]
@edwarnicke 
Currently @d-uzlov is working on the whole NSM scalability testing. Should we start working on components scalability testing in the closest future or is it a low priority task and so it should be postponed for some time after?
